Artifact Protocol + Negotiation Cards

turingcompl33t opened the pull request:
Resolves #151.

This PR introduces the MLTE artifact protocol; a standardization of many of the practices we have adopted in the codebase over time for managing the interaction between the source client (the MLTE library itself) the web client (the MLTE web frontend) and persistent storage in the MLTE artifact store. Specifically, this PR introduces the following changes:

- It introduces the MLTE artifact protocol, which will be the new mechanism to ensure consistent interfaces across all MLTE artifacts
- It introduces the implementation of the negotiation card artifact; this is the first artifact to implement the "artifact protocol" and demonstrates the procedure for implementing it
- It introduces `namespaces` to the MLTE context; this is an important organizational mechanism that was missing previously
- It introduces a new implementation of the MLTE `Store` type, which is the in-memory representation of the MLTE artifact store. Previously this type was bound to the underlying storage mechanism; now it is a generic "client" for the underlying storage mechanism that makes interaction with the store consistent
- It introduces a new implementation of the artifact store web backend; the API for managing organizational elements (namespaces, models, model versions) as well as for managing artifacts has been revised and standardized

This PR also introduces some technical debt. I have removed the existing hooks from various parts of the codebase that manage the interaction between artifacts and the artifact store. These hooks will need to be replaced as we port these artifacts to the protocol. Furthermore, I had to disable unit tests for this functionality for the affected artifacts.

sebastian-echeverria commented on src/mlte/store/store.py:
```diff
@@ -0,0 +1,325 @@
+            "Cannot invoke method on abstract StoreSession."
+        )
+
+    def write_negotiation_card(
```
I am a little confused about the NegotiationCard being here. Are all artifacts going to have methods inside StoreSession? If so, this won't scale, or at best it is going to get really long. Is that the idea, for all artifacts to have 3 methods in StoreSession? It doesn't seem appropriate . Or is NegotiationCard special in some way that requires it to be in here? It is not part of the session, it looks like just another artifact to me.

turingcompl33t replied:
Yes, the idea in the current implementation is to have 3 methods within `StoreSession` (and each underlying implementation, actually) for each artifact type.

Originally, I intended the interface to include only a 3 methods for artifacts total, by operating on an abstract `ArtifactModel` instead of the individual artifact types themselves. I will revisit this to determine if I can make it more generic and thus scalable.

turingcompl33t replied:
Refactored this implementation. See updates.

sebastian-echeverria commented on src/mlte/store/underlying/http.py:
```diff
@@ -0,0 +1,281 @@
+    # Negotiation Card
+    # -------------------------------------------------------------------------
+
+    def write_negotiation_card(
+        self,
+        namespace_id: str,
+        model_id: str,
+        version_id: str,
+        artifact: NegotiationCardModel,
+    ) -> NegotiationCardModel:
+        url = f"{_url(self.url, namespace_id, model_id, version_id)}/negotiation-card"
+        res = self.client.post(url, json=artifact.dict())
+        raise_for_response(res)
+
+        return NegotiationCardModel(**res.json())
```
Related to my previous comment, do we really need write/read/delete methods for each artifact separately? Can't we have a generic write/read/delete artifact method, that maybe. receives the model type and gets additional info, like URL, as needed? This looks like it will require a lot of almost-the-same code for each artifact, with only very minor difference (basically the model type and the actual URL. I like trying to avoid duplicating samey code if possible to make maintenance easier.

turingcompl33t replied:
See refactor.

sebastian-echeverria commented on src/mlte/web/store/api/endpoints/metadata.py:
Nice to see this shaping up

turingcompl33t replied:
Agreed!

sebastian-echeverria reviewed (commented):
Ok, I was able to do a quick review of the PR, not as in depth as I'd like to, but I wanted to send you some feedback before going on PTO.  I like the overall design, the use of Models to describe and handle serialization looks nice, the Artifact abstraction looks good as well, the changes to Context to be more self contained are awesome, and the way Session is handled looks fine overall as well. The direction the Store is taking seems good too. 

I had a couple of questions I mentioned below, mostly to do with duplicating samey code vs having named methods for each artifact write/read/delete. You could have common methods that do most of the code and still have named methods for each specific artifact, that would be one improvement.  But I am also worried some things related to session and store may end up being too large, with 3 methods for each artifact. You can review that and we can discuss that once I get back.

Also, as noted in one comment, you seem to be accidentally undoing one of the changes that Alex added to integrate the frontend.

turingcompl33t commented on card.json:
This was a mistake. Thanks for the catch. Removed.

turingcompl33t commented on src/mlte/cli/cli.py:
This was a mishandled conflict. I added this back in.

turingcompl33t commented:
@sebastian-echeverria I refactored the `Store` interface to generically accept an `ArtifactModel` rather than a distinct model for each artifact type. This should make it simpler to scale to new artifact types. One weakness of the design I introduced is that artifacts must now have a unique name _across_ artifact types. However, this is relatively minor from an implementation perspective, so we can continue to push forward with this and address that if necessary.

sebastian-echeverria commented on src/mlte/store/store.py:
Improvements look good too reduce duplication

turingcompl33t replied:
Thanks!

sebastian-echeverria commented on src/mlte/web/store/api/endpoints/artifact.py:
```diff
@@ -0,0 +1,139 @@
+    namespace_id: str, model_id: str, version_id: str, artifact_id: str
+) -> ArtifactModel:
+    """
+    Read a negotiation card.
```
Comment is outdated

turingcompl33t replied:
Good catch, got this one.

sebastian-echeverria commented on src/mlte/web/store/api/endpoints/artifact.py:
```diff
@@ -0,0 +1,139 @@
+    namespace_id: str, model_id: str, version_id: str, query: Query
+) -> list[ArtifactModel]:
+    """
+    Read a negotiation card.
```
Comment is outdated

turingcompl33t replied:
Got this one.

sebastian-echeverria commented on src/mlte/web/store/api/endpoints/artifact.py:
```diff
@@ -0,0 +1,139 @@
+    namespace_id: str, model_id: str, version_id: str, artifact_id: str
+) -> ArtifactModel:
+    """
+    Delete a negotiation card.
```
Same

turingcompl33t replied:
Got it.

sebastian-echeverria commented on .github/ISSUE_TEMPLATE/artifact-protocol-port.md:
```diff
@@ -0,0 +1,28 @@
+	- `InMemoryStore`
+	- `RemoteHttpStore`
+- Add relevant endpoints to the web store implementation
+- Add unit tests to the web store implementation
```
Is any of this still needed after the changes to use the more generic ArtifactModel and the read/write/delete_artifact methods?

turingcompl33t replied:
Nope, much of the extra boilerplate for the port goes away now. I've updated the issues accordingly.

sebastian-echeverria reviewed (changes requested):
Ok, not sure if you finished doing the refactoring you wanted to do, but since I just got back from PTO I reviewed your changes and added some comments. The refactoring looks awesome and it looks like it will save a lot of samey code when adding new artifacts! I understand the weakness about artifact ids having to be unique among all artifact types, but I agree it is not a terrible issue; Github for example does something similar between issues and PRs and other types of things. If there are any issues from this we can handle them later. I had some minor comments, mostly about some things that seems to be a bit out of date and still referencing Negotiation Card instead of the more generic structures/name. Thanks!

turingcompl33t commented on .github/ISSUE_TEMPLATE/artifact-protocol-port.md:
Thanks!

turingcompl33t commented on src/mlte/store/query.py:
Thanks!

turingcompl33t commented on src/mlte/artifact/model.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class ArtifactHeaderModel(BaseModel):
+    """The NegotiationCardHeaderModel contains the common metadata for all artifacts."""
```
Good catch, updated.

turingcompl33t commented on src/mlte/artifact/model.py:
```diff
@@ -0,0 +1,37 @@
+    header: ArtifactHeaderModel
+    """The artifact header."""
+
+    body: NegotiationCardModel
```
This is confusing. My plan here is to make the body a `Union` over all of the various artifact types, so this will be:

```python
header: ArtifactHeaderModel

body: Union[SpecModel, NegotiationCardModel, ...]
```

The `Union` annotation just won't allow a `Union` with a single type, and the `NegotiationCardModel` is the only one implemented at present.

turingcompl33t replied:
Roger, updated this with a comment.

turingcompl33t commented:
Just one more comment unresolved from your previous review @sebastian-echeverria. If you are happy with this one, we can resolve and merge and start working on the next steps together (and bring in @sei-aderr as well).

sebastian-echeverria commented:
I am fine with all the answers, maybe you could just add a comment as I mentioned above for the Union plan, and with that we should be fine to merge and start the next steps.

turingcompl33t commented:
Great, updated that implementation with a comment. Will merge this after the pipelines pass.

sebastian-echeverria reviewed (approved):
Looks good!

sebastian-echeverria reviewed (approved):
Done

sebastian-echeverria reviewed (approved):
Approved

sebastian-echeverria reviewed (approved):
Really approved!
